aiida/orm/querybuilder.py

```diff
@@ -491,25 +506,82 @@ def __init__(
+        return self.as_dict()
+
+    @classmethod
+    def from_dict(cls, dct: QueryDict) -> 'QueryBuilder':
+        """Create an instance from a dictionary representation of the query."""
+        return cls(**dct)  # type: ignore[arg-type]
+
+    @staticmethod
+    def _compile_query(query: 'Query', literal_binds: bool = False) -> 'SQLCompiler':
+        """Compile the query to the SQL executable.
+
+        :params literal_binds: Inline bound parameters (this is normally handled by the Python DBAPI).
         """
-        from aiida.manage.configuration import get_config
+        dialect = query.session.bind.dialect
 
-        config = get_config()
-        engine = config.current_profile.database_engine
+        class _Compiler(dialect.statement_compiler):  # type: ignore[name-defined]
+            """Override the compiler with additional literal value renderers."""
 
-        if engine.startswith('mysql'):
-            from sqlalchemy.dialects import mysql as mydialect
-        elif engine.startswith('postgre'):
-            from sqlalchemy.dialects import postgresql as mydialect
-        else:
-            raise ConfigurationError(f'Unknown DB engine: {engine}')
+            def render_literal_value(self, value, type_):
+                """Render the value of a bind parameter as a quoted literal.
 
-        que = self.get_query()
-        return str(que.statement.compile(compile_kwargs={'literal_binds': True}, dialect=mydialect.dialect()))
+                See https://www.postgresql.org/docs/current/functions-json.html for serialisation specs
+                """
+                try:
+                    return super().render_literal_value(value, type_)
+                except NotImplementedError:
+                    if isinstance(value, list):
+                        values = ','.join(self.render_literal_value(item, type_) for item in value)
+                        return f"'[{values}]'"
+                    if isinstance(value, int):
+                        return str(value)
+                    if isinstance(value, (str, date, datetime, timedelta)):
+                        escaped = str(value).replace('"', '\\"')
+                        return f'"{escaped}"'
+                    raise
+
+        return _Compiler(dialect, query.statement, compile_kwargs=dict(literal_binds=literal_binds))
+
+    def as_sql(self, inline: bool = False) -> str:
+        """Convert the query to an SQL string representation.
+
+        .. warning::
+
+            This method should be used for debugging puposes only,
+            since normally sqlalchemy will handle this process internally.
+
+        :params inline: Inline bound parameters (this is normally handled by the Python DBAPI).
+        """
+        compiled = self._compile_query(self.get_query(), literal_binds=inline)
+        if inline:
+            return compiled.string + '\n'
+        return f'{compiled.string!r} % {compiled.params!r}\n'
+
+    def __str__(self) -> str:
```

Comment:
Wouldn't this be more the `__repr__` functionality? Although I can't think of the essentials to show for the string representation?

Reply:
so with this PR:

```python
        qb = orm.QueryBuilder().append(orm.Data, project=['id', 'uuid']).order_by({orm.Data: 'id'})
        print(repr(qb))
        print(qb)
```

goes to

```python
<aiida.orm.querybuilder.QueryBuilder object at 0x7fbd0aa64410>

QueryBuilder(path=[{'entity_type': 'data.Data.', 'tag': 'Data_1', 'joining_keyword': None, 'joining_value': None, 'edge_tag': None, 'outerjoin': False}], filters={'Data_1': {'node_type': {'like': 'data.%'}}}, project={'Data_1': [{'id': {}}, {'uuid': {}}]}, order_by=[{'Data_1': [{'id': {'order': 'asc'}}]}], limit=None, offset=None)
```

Personally I think this makes sense, since the `__repr__` uniquely identifies the object, and the string `__str__` provides a more verbose definition of it (and the bonus is you can literally copy/paste it to recreate the instance)

Reply:
I wouldn't say the `__repr__` defines the instance in a usable unique way, other than referencing the point in memory. Again, I'm more for the `__repr__` string being used for this exact purpose (recreating an instance) - this is the way it is used in the built-in Python libs as well - but that said this is already discussed extensively in #3087.

Reply:
yeh lets leave this discussion to that issue lol; IMO this is exactly what the repr and str should provide, otherwise they would just be identical.
Anyhow, I think its certainly better than the current situation, of showing the raw SQL to users (which can also except)

Reply:
Sure - but again you're introducing another way of string representing AiiDA entities in the code base now. Maybe do it, but make an explicit note of it the general issue - although a future PR for #3087 should already take *all* AiiDA entities into account anyhow ... so yeah. Fine, I guess? 😅

Reply:
The QueryBuilder is not an aiida entity though, i.e. it does not relate to an ORM object, and its string representation was already different to anything else in aiida; so its not introducing another way, its just changing the current way

Reply:
Hmm, I'm fuzzy on that definition exactly myself. But if you consider everything that is stored as a row in a PostgreSQL table an entity, I'll agree. But I'm not entirely sure that's the definition. But maybe it should be.

Reply:
In fact might just make this both the repr and str

Reply:
yeh I've just made the repr and str the same, the basically what happens for all the python built-in stuff